### PR TITLE
Add custom instruction_proposer support to GEPA with multimodal (dspy.Image) handling

### PR DESCRIPTION
**Pull request by andressrg**
Closes https://github.com/stanfordnlp/dspy/issues/8690

# Add support for custom instruction proposers in GEPA with multimodal handling.

As proposed by @LakshyAAAgrawal, this PR makes GEPA’s instruction-proposal step pluggable. Users can supply a custom instruction proposer, enabling GEPA to handle dspy.Image and other custom types.

## Key Features

- **Custom Instruction Proposer Protocol**: New `InstructionProposerProtocol` for implementing custom instruction generation
- **MultiModalInstructionProposer**: Built-in proposer that handles multimodal inputs (text + images) during optimization
- **Structured Image Handling**: Preserves `dspy.Image` objects without serialization in reflection LM calls
- **Backward Compatibility**: Existing GEPA usage continues to work unchanged

### For multi-modal use cases

```python
import dspy
from dspy.teleprompt.gepa.instruction_proposal import MultiModalInstructionProposer

# Configure your language models
lm = dspy.LM(model='gpt-4o-mini')
reflection_lm = dspy.LM(model='gpt-4o', temperature=1.0, max_tokens=32000)

# Create a program that uses images
program = dspy.Predict("image: dspy.Image -> label: str")

# Set up GEPA with custom instruction proposer
gepa = dspy.GEPA(
    metric=your_metric_function,
    max_metric_calls=50,
    reflection_lm=reflection_lm,
    instruction_proposer=MultiModalInstructionProposer()  # Enable multimodal support
)

# Compile with image data
compiled_program = gepa.compile(program, trainset=image_trainset, valset=image_valset)
```

### Advanced: Custom Instruction Proposer

You can implement your own instruction proposer by following the protocol:

```python
from dspy.teleprompt.gepa.instruction_proposal import InstructionProposerProtocol

class MyCustomProposer:
    def __call__(self, current_instruction: str, reflective_dataset: list[dict]) -> str:
        # Your custom logic here
        # reflective_dataset contains examples with inputs, outputs, and feedback
        # May include dspy.Image objects in inputs
        
        # Analyze the feedback and generate improved instruction
        improved_instruction = analyze_and_improve(current_instruction, reflective_dataset)
        return improved_instruction

# Use with GEPA
gepa = dspy.GEPA(
    metric=your_metric,
    max_metric_calls=50,
    reflection_lm=reflection_lm,
    instruction_proposer=MyCustomProposer()
)
```

## Backward Compatibility

This change is fully backward compatible. Existing GEPA code will continue to work without modification:

```python
# This still works exactly as before
gepa = dspy.GEPA(metric=my_metric, max_metric_calls=50, reflection_lm=reflection_lm)
compiled = gepa.compile(program, trainset, valset)
```

## Files Changed

- `dspy/teleprompt/gepa/gepa.py`: Add `instruction_proposer` parameter
- `dspy/teleprompt/gepa/gepa_utils.py`: Integrate custom proposer logic and preserve image objects
- `dspy/teleprompt/gepa/instruction_proposal.py`: New multimodal instruction proposer implementation
- `tests/teleprompt/test_gepa_instruction_proposer.py`

This enables GEPA optimization to work with visual content while maintaining proper image handling throughout the reflection and instruction proposal process.

### CODE REVIEW

File: dspy/teleprompt/gepa/gepa.py [hunk 1]
```diff
@@ -282,6 +283,8 @@ def __init__(
         track_best_outputs: bool = False,
         # Reproducibility
         seed: int | None = 0,
+        # Custom instruction proposer
```
Comment (LakshyAAAgrawal) [inline dspy/teleprompt/gepa/gepa.py]: Could you move this up to # Reflection based configuration section (after line 269)

Comment (LakshyAAAgrawal) [reply dspy/teleprompt/gepa/gepa.py]: And make a corresponding edit to the docstring.

File: dspy/teleprompt/gepa/gepa_utils.py [hunk 2]
```diff
@@ -1,13 +1,18 @@
...
 from gepa import EvaluationBatch, GEPAAdapter
 
+import dspy
 from dspy.adapters.chat_adapter import ChatAdapter
 from dspy.adapters.types import History
+from dspy.adapters.types.base_type import Type
 from dspy.evaluate import Evaluate
 from dspy.primitives import Example, Prediction
+
+if TYPE_CHECKING:
```
Comment (LakshyAAAgrawal) [inline dspy/teleprompt/gepa/gepa_utils.py]: We perhaps don't need the `if` condition here?

File: dspy/teleprompt/gepa/gepa_utils.py [hunk 3]
```diff
@@ -66,6 +73,35 @@ def __init__(
         self.num_threads = num_threads
         self.add_format_failure_as_feedback = add_format_failure_as_feedback
         self.rng = rng or random.Random(0)
+        self.reflection_lm = reflection_lm
+        self.custom_instruction_proposer = custom_instruction_proposer
+
+        if self.custom_instruction_proposer is not None:
```
Comment (LakshyAAAgrawal) [inline dspy/teleprompt/gepa/gepa_utils.py:77]: There should be an assert here, that ensures whenever `self.custom_instruction_proposer is not None`, then `assert self.reflection_lm is not None` as well.

File: dspy/teleprompt/gepa/gepa_utils.py [hunk 4]
```diff
@@ -227,29 +268,3 @@ def make_reflective_dataset(self, candidate, eval_batch, components_to_update):
             raise Exception("No valid predictions found for any module.")
 
         return ret_d
-
-    # TODO: The current DSPyAdapter implementation uses the GEPA default propose_new_texts.
```
Comment (LakshyAAAgrawal) [inline dspy/teleprompt/gepa/gepa_utils.py]: I recommend retaining this part, as it acts as documentation for future readers.

File: dspy/teleprompt/gepa/instruction_proposal.py [hunk 5]
```diff
@@ -0,0 +1,280 @@
+from typing import Any, Protocol
+
+import dspy
+from dspy.adapters.types.base_type import Type
+
+
+class InstructionProposerProtocol(Protocol):
```
Comment (LakshyAAAgrawal) [inline dspy/teleprompt/gepa/instruction_proposal.py]: GEPA already has a protocol very similar to this protocol: https://github.com/gepa-ai/gepa/blob/main/src/gepa/core/adapter.py#L29. I am wondering if having a separate protocol supporting very similar (same?) functionality will lead to a drift.

The important of `components_to_update` parameter is that: While currently GEPA only selects one prompt to update at a time, there are many DSPy programs where prompts of 2 different predictors could be so tightly coupled that they need to be generated together in a single LLM call. While this is currently not supported, we should keep the doors open for supporting this (in the very near future)!

Comment (LakshyAAAgrawal) [reply dspy/teleprompt/gepa/instruction_proposal.py]: I recommend that for now, we can just import GEPA's protocol here, and define a default outer loop, which proposes individual prompts exactly as you do below. In future, somebody could simply override the default loop function, to do the kinds of customization I talk about above.

File: dspy/teleprompt/gepa/instruction_proposal.py [hunk 6]
```diff
@@ -0,0 +1,280 @@
...
+            reflective_dataset: List of examples with inputs, outputs, and feedback
+                               May contain dspy.Image objects or other custom types
+
+        Returns:
+            str: Improved instruction text
+        """
+        ...
+
+
+class GenerateEnhancedMultimodalInstructionFromFeedback(dspy.Signature):
```
Comment (LakshyAAAgrawal) [inline dspy/teleprompt/gepa/instruction_proposal.py:9]: These proposers are AMAZING! Thank you very much

**Review by LakshyAAAgrawal — CHANGES_REQUESTED** [review]

**Comment by LakshyAAAgrawal** [comment]
Dear @andressrg,

Thank you so much for kindly contributing the custom instruction proposal to GEPA. The implementation is great. I have added a few suggestions. Overall I think this is good to merge.

File: dspy/teleprompt/gepa/gepa.py (hunk 1 above)
Comment (jzabroski) [reply dspy/teleprompt/gepa/gepa.py]: `# Reflection based configuration` should also probably say `# Reflection configuration` so that the docstring and code are the same and the code comment functions as a backreference.

Comment (LakshyAAAgrawal) [reply dspy/teleprompt/gepa/gepa.py]: Good call @jzabroski , thanks!

File: dspy/teleprompt/gepa/instruction_proposal.py (hunk 5 above)
Comment (andressrg) [reply dspy/teleprompt/gepa/instruction_proposal.py]: Great point!

Your proposal is a lot more powerful.

Done.

File: dspy/teleprompt/gepa/gepa.py (hunk 1 above)
Comment (andressrg) [reply dspy/teleprompt/gepa/gepa.py]: For sure!

Updated

File: dspy/teleprompt/gepa/gepa_utils.py (hunk 2 above)
Comment (andressrg) [reply dspy/teleprompt/gepa/gepa_utils.py]: correct. Removed.

File: dspy/teleprompt/gepa/gepa_utils.py (hunk 3 above)
Comment (andressrg) [reply dspy/teleprompt/gepa/gepa_utils.py:77]: good point. Added 👍

File: dspy/teleprompt/gepa/instruction_proposal.py (hunk 6 above)
Comment (andressrg) [reply dspy/teleprompt/gepa/instruction_proposal.py:9]: Glad you like them 🚀 🚀 🚀

File: dspy/teleprompt/gepa/gepa_utils.py (hunk 4 above)
Comment (andressrg) [reply dspy/teleprompt/gepa/gepa_utils.py]: good point. Added back.

**Comment by andressrg** [comment]
@LakshyAAAgrawal Thank you for the detailed feedback! All comments have been addressed:

- Refactored to use GEPA's native ProposalFn protocol instead of custom InstructionProposerProtocol
- Moved instruction_proposer to Reflection configuration section and updated docstring
- Added assertion for reflection_lm when using custom proposer
- Kept the TODO comment for documentation purposes

The new ProposalFn-based API is indeed much more powerful and extensible. It opens the door for multi-component instruction updates in the future as you mentioned. Great suggestion.

File: dspy/teleprompt/gepa/gepa.py [hunk 7]
```diff
@@ -211,6 +211,13 @@ def metric(
             a strong reflection model. Consider using `dspy.LM(model='gpt-5', temperature=1.0, max_tokens=32000)` 
             for optimal performance.
         skip_perfect_score: Whether to skip examples with perfect scores during reflection. Default is True.
+        instruction_proposer: Optional custom instruction proposer implementing GEPA's ProposalFn protocol. 
+            If provided, GEPA will use this custom proposer instead of its default instruction proposal
```
Comment (LakshyAAAgrawal) [inline dspy/teleprompt/gepa/gepa.py:213]: Can you include some link to the `default instruction proposal` so that people wanting to implement their own logic have a reference point? Also some kind of link to `MultiModalInstructionProposer() from dspy.teleprompt.gepa.instruction_proposal`.

File: dspy/teleprompt/gepa/gepa.py [hunk 8]
```diff
@@ -10,9 +10,9 @@
 
 if TYPE_CHECKING:
```
Comment (LakshyAAAgrawal) [inline dspy/teleprompt/gepa/gepa.py]: TYPE_CHECKING remove as earlier

File: dspy/teleprompt/gepa/gepa_utils.py [hunk 9]
```diff
@@ -77,29 +75,22 @@ def __init__(
...
-                    dataset_with_feedback = reflective_dataset[name]
-
-                    with dspy.context(lm=self.reflection_lm):
-                        new_texts[name] = proposer(
-                            current_instruction=current_instruction,
-                            reflective_dataset=dataset_with_feedback
-                        )
-
-                return new_texts
+                with dspy.context(lm=self.reflection_lm):
```
Comment (LakshyAAAgrawal) [inline dspy/teleprompt/gepa/gepa_utils.py]: Another flexibility we can provide here is, instead of mandating that the reflection_lm be provided and using `dspy.context` here, which essentially forces me to pick one reflection_lm and stick with it through optimization,

we can just let the user provide a custom_instruction_proposer, and call it. The custom_instruction_proposer takes care of what the reflection_lm is. If reflection_lm is provided in addition, then we use it, otherwise it is not mandatory.

**Review by LakshyAAAgrawal — CHANGES_REQUESTED** [review]
Left a few more suggestions, I think this is the final round!

File: dspy/teleprompt/gepa/gepa.py (hunk 8 above)
Comment (andressrg) [reply dspy/teleprompt/gepa/gepa.py]: Sure

Done

File: dspy/teleprompt/gepa/gepa.py (hunk 7 above)
Comment (andressrg) [reply dspy/teleprompt/gepa/gepa.py:213]: I'm not sure how to add links here. I've just added this:

```
Use `MultiModalInstructionProposer()` from `dspy.teleprompt.gepa.instruction_proposal` 
for handling visual content. If None (default), GEPA uses its built-in text-optimized 
proposer (see `gepa.strategies.instruction_proposal.InstructionProposalSignature` 
for reference implementation).
```

Is this what you were proposing? or is there a way to add hrefs here that I might not be familiar with? 🤔

File: dspy/teleprompt/gepa/gepa_utils.py (hunk 9 above)
Comment (andressrg) [reply dspy/teleprompt/gepa/gepa_utils.py]: I like the idea!

Just made this change:

```
        if self.custom_instruction_proposer is not None:
            # We are only overriding the propose_new_texts method when a custom
            # instruction proposer is provided. Otherwise, we use the GEPA
            # default propose_new_texts.

            def custom_propose_new_texts(
                candidate: dict[str, str],
                reflective_dataset: dict[str, list[dict[str, Any]]],
                components_to_update: list[str]
            ) -> dict[str, str]:
                if self.reflection_lm is not None:
                    with dspy.context(lm=self.reflection_lm):
                        return self.custom_instruction_proposer(
                            candidate=candidate,
                            reflective_dataset=reflective_dataset,
                            components_to_update=components_to_update
                        )
                else:
                    return self.custom_instruction_proposer(
                        candidate=candidate,
                        reflective_dataset=reflective_dataset,
                        components_to_update=components_to_update
                    )

            self.propose_new_texts = custom_propose_new_texts
```

Now, the core `gepa/appi.py` does require the user to provide a reflection_lm:

https://github.com/gepa-ai/gepa/blob/2cf10c79125533af345ffbd48497005dd1f68ee9/src/gepa/api.py#L127

Which means that in order to have true flexibility, the core gepa library needs to be updated as well I believe.

**Comment by andressrg** [comment]
Updated!

Left a question, and one comment, to make sure I understand your feedback correctly.

Thanks

File: dspy/teleprompt/gepa/gepa.py (hunk 7 above)
Comment (LakshyAAAgrawal) [reply dspy/teleprompt/gepa/gepa.py:213]: I think this is markdown so \[]() should work. I think you can feel free to ignore this as nit for now, otherwise, just include a markdown link to the source location. Won't hold back merge for this!

File: dspy/teleprompt/gepa/gepa.py [hunk 10]
```diff
@@ -309,11 +317,11 @@ def __init__(
         self.max_full_evals = max_full_evals
         self.max_metric_calls = max_metric_calls
 
-        # Reflection based configuration
+        # Reflection configuration
         self.reflection_minibatch_size = reflection_minibatch_size
         self.candidate_selection_strategy = candidate_selection_strategy
-        # self.reflection_lm = reflection_lm
         assert reflection_lm is not None, "GEPA requires a reflection language model to be provided. Typically, you can use `dspy.LM(model='gpt-5', temperature=1.0, max_tokens=32000)` to get a good reflection model. Reflection LM is used by GEPA to reflect on the behavior of the program and propose new instructions, and will benefit from a strong model."
```
Comment (LakshyAAAgrawal) [inline dspy/teleprompt/gepa/gepa.py]: This should be modified to say "assert custom_proposer is not None or reflection_lm is not None"

**Review by LakshyAAAgrawal — CHANGES_REQUESTED** [review]

File: dspy/teleprompt/gepa/gepa_utils.py (hunk 9 above)
Comment (LakshyAAAgrawal) [reply dspy/teleprompt/gepa/gepa_utils.py]: Good catch. Let me update that!

Comment (LakshyAAAgrawal) [reply dspy/teleprompt/gepa/gepa_utils.py]: Can you take a look at: https://github.com/gepa-ai/gepa/pull/50

I think this should fix it. Can you check?

Comment (andressrg) [reply dspy/teleprompt/gepa/gepa_utils.py]: Nice!. Checking now

Comment (andressrg) [reply dspy/teleprompt/gepa/gepa_utils.py]: It does fix it 🚀 

Now dspy.GEPA can support reflection_lm=None

**Comment by LakshyAAAgrawal** [comment]
Released gepa v0.0.8 (with the disabled assert). You can update the pyproject.toml to reflect the change, and test against that!

If possible, it would be great to include a unit test.

**Comment by andressrg** [comment]
Great! 

On it.

File: dspy/teleprompt/gepa/gepa.py (hunk 10 above)
Comment (andressrg) [reply dspy/teleprompt/gepa/gepa.py]: sure. Done

File: uv.lock [hunk 11]
```diff
@@ -1,5 +1,5 @@
 version = 1
-revision = 2
+revision = 3
```
Comment (andressrg) [inline uv.lock]: @LakshyAAAgrawal I'm aware my uv (version 0.8.14) updated the revision here.

Let me know if you prefer me to revert it.

**Comment by andressrg** [comment]
@LakshyAAAgrawal Changes ready.

Also added a test to validate the system works as expected if you provide an instruction_proposer and reflection_lm=None.

File: dspy/teleprompt/gepa/gepa.py [hunk 12]
```diff
@@ -210,6 +209,14 @@ def metric(
             a strong reflection model. Consider using `dspy.LM(model='gpt-5', temperature=1.0, max_tokens=32000)` 
             for optimal performance.
         skip_perfect_score: Whether to skip examples with perfect scores during reflection. Default is True.
+        instruction_proposer: Optional custom instruction proposer implementing GEPA's ProposalFn protocol.
```
Comment (LakshyAAAgrawal) [inline dspy/teleprompt/gepa/gepa.py:212]: Let's add a note here, saying that "if instruction_proposer is provided and reflection_lm is provided, the instruction_proposer is called in context of the reflection_lm. However, it is not mandatory to provide reflection_lm and instruction_proposer can implement custom logic including invocation of custom LLMs." Feel free to improve upon this!

File: uv.lock (hunk 11 above)
Comment (LakshyAAAgrawal) [reply uv.lock]: Yes, let's revert it

File: uv.lock [hunk 13]
```diff
@@ -664,7 +664,7 @@ wheels = [
 
 [[package]]
 name = "dspy"
-version = "3.0.0"
+version = "3.0.1"
```
Comment (LakshyAAAgrawal) [inline uv.lock]: Should this be 3.0.3?

Comment (LakshyAAAgrawal) [reply uv.lock]: Probably it would be good to merge back/rebase from main and then merge this after.

File: dspy/teleprompt/gepa/gepa.py (hunk 12 above)
Comment (andressrg) [reply dspy/teleprompt/gepa/gepa.py:212]: sure. On it

Comment (andressrg) [reply dspy/teleprompt/gepa/gepa.py:212]: Added

File: uv.lock (hunk 13 above)
Comment (andressrg) [reply uv.lock]: good catch!

Rebased and ran uv lock again.

**Comment by andressrg** [comment]
Feedback addressed.

I like the feedback! We want to land a polished PR 🚀

**Review by LakshyAAAgrawal — APPROVED** [review]

**Comment by LakshyAAAgrawal** [comment]
@andressrg, Thank you very very much for this super amazing PR, and then also going through so many rounds of updates. The PR is very high quality, and adds a very useful feature to the dspy.GEPA optimizer.

Thank you very much for your contribution!

CC: @okhat

**Comment by andressrg** [comment]
Thank you very for the guidance and the feedback!

GEPA is amazing and I'm looking forward to all the vision use-cases that can be optimized by leveraging it!
